[12.0][ADD] hr_timesheet_sheet_prefill_multi

carmenbianca opened the pull request:
## Description

Be able to both sort and repeat default projects for timesheet sheets.

I'm pretty content with how this module turned out! It is as hacky as it is elegant.

To save myself from having to track two PRs, we can just merge this with `merge patch`. It will also bump `hr_timesheet_sheet_prefill_multi`, but that doesn't really matter. Version numbers are arbitrary, anyway.

## Odoo task (if applicable)

T12533

## Checklist before approval

- [x] Tests are present (or not needed).
- [x] Credits/copyright have been changed correctly.
- [x] (If a new module) Moving this to OCA has been considered.

carmenbianca committed:
[FIX] hr_timesheet_sheet_prefill: Define table name of M2m relation

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] hr_timesheet_sheet_prefill: Add @api.model to methods

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[REF] hr_timesheet_sheet_prefill: Add hook for new module

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[ADD] hr_timesheet_sheet_prefill_multi

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] hr_timesheet_sheet_prefill_multi: Mark as rebel

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

huguesdk commented on hr_timesheet_sheet_prefill_multi/__manifest__.py:
```diff
@@ -0,0 +1,23 @@
+
+{
+    "name": "Timesheet Sheet prefill with duplicates",
+    "summary": """
+        Allow duplicates in prefill templates.""",
```
```suggestion
    "summary": "Allow duplicates in prefill templates",
```

huguesdk commented on hr_timesheet_sheet_prefill_multi/__manifest__.py:
```diff
@@ -0,0 +1,23 @@
+    "author": "Coop IT Easy SC, Odoo Community Association (OCA)",
+    "maintainers": ["carmenbianca"],
+    "license": "AGPL-3",
+    "application": False,
```
```suggestion
```

huguesdk commented on hr_timesheet_sheet_prefill_multi/views/hr_employee_views.xml:
```diff
@@ -0,0 +1,32 @@
+        <field name="arch" type="xml">
+            <field name="project_ids" position="after">
+                <field name="timesheet_prefill_ids">
+                    <tree string="Prefill Projects" editable="bottom">
```
```suggestion
                    <tree editable="bottom">
```
`string` is deprecated since version 8.0.

huguesdk commented on hr_timesheet_sheet_prefill_multi/views/hr_employee_views.xml:
```diff
@@ -0,0 +1,32 @@
+        />
+        <field name="arch" type="xml">
+            <field name="project_ids" position="after">
+                <field name="timesheet_prefill_ids">
```
```suggestion
                <field name="timesheet_prefill_ids" nolabel="1">
```
since it’s the only element in the “Default Projects” section, or:
```suggestion
                <field name="timesheet_prefill_ids" string="Projects">
```
so that it appears as the other field, since it has the same use.

huguesdk reviewed (changes requested):
a few cosmetics remarks, otherwise lgtm! well done, @carmenbianca! 👍

huguesdk commented on hr_timesheet_sheet_prefill_multi/models/hr_employee.py:
```diff
@@ -0,0 +1,25 @@
+        # to the sequence of the prefill records.
+        projects = self.env["project.project"].browse()
+        for prefill in self.timesheet_prefill_ids.sorted():
+            projects += prefill.project_project_id
```
in order to stay compatible with the way `hr_timesheet_sheet_prefill` works, this should check whether the project is active and only add it if it’s the case. inactive projects should not appear, as they will cause an error while trying to create timesheet entries.

carmenbianca replied:
Implemented this in `hr_timesheet_sheet_prefill`, and added a domain to make sure that users can't add inactive projects (although this is not enforced as a restraint).

carmenbianca committed:
[FIX] hr_timesheet_sheet_prefill_multi: Cosmestic fixes

A very small patchset to clean up the code.

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[IMP] hr_timesheet_sheet_prefill[_multi]: Only use active projects

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca commented:
@huguesdk done

huguesdk committed:
[IMP] simplify filtering out of inactive projects

use a related field on hr_timesheet.sheet.prefill, which solves the
problem elegantly and also ensures that inactive projects are hidden in
the prefill projects of employees.

huguesdk commented:
i added a commit to simplify the filtering out of inactive projects. you may want to squash it with the previous one.

carmenbianca commented:
@huguesdk LGTM. I think it's OK to keep as separate commit.

huguesdk committed:
[REL] hr_timesheet_sheet_prefill 12.0.1.1.0

huguesdk commented:
/ocabot merge nobump
